Load data from the OpenAQ data archive on AWS Open Data

zmoon opened the pull request:
* [ ] add imports
* [ ] more tests (e.g. loading data for more than one site)
* [ ] wide format conversion (maybe use the func from the v3 web API once #198 is merged)
* [ ] it would be nice to have more site metadata, which `get_locations` from the v3 web API reader could provide

zmoon committed:
Initial file reader

zmoon committed:
Initial add_data

very slow if going for all locations though

zmoon committed:
Seems like using exists is a bit faster

in the case of one month for two locs...

zmoon committed:
Get loc IDs from directory structure

zmoon committed:
By provider

testing with aqdc, 2017-07, takes a few min to find files,
though after the first date query the next ones seem much faster

found almost 4k files, loaded in a few min with 4 proc

zmoon committed:
API consistency

zmoon committed:
By country

tried mx (Mexico), which seems to have had just 40 unique locs
in 2017-07, still took 5-ish minutes to discover the 700-ish files

zmoon committed:
Funcs to help load data from a list of sites

without pre-checking if the files exist

zmoon committed:
Try discovering all site-days

took 18 hours to find 17_455_910 site-days,
so wouldn't want to do it often

.csv.gz at 41 MB

zmoon committed:
Tweak name

zmoon committed:
Some name tweaks

zmoon committed:
unit -> units

this is the standard name
going this way with the v3 API reader too

zmoon committed:
More discovery options

zmoon committed:
Add naive build-url option for add-data

zmoon committed:
Return df from get-locations showing country and provider

zmoon committed:
Add some tests

zmoon committed:
Test add-data for one site-day

zmoon commented on monetio/obs/openaq_aws.py:
```diff
@@ -0,0 +1,511 @@
+
+https://registry.opendata.aws/openaq/
+
+https://docs.openaq.org/docs/accessing-openaq-archive-data
```
```suggestion
https://docs.openaq.org/aws/about
```
it was changed

zmoon committed:
Fix link

bbakernoaa commented:
@zmoon is this ok to merge?

zmoon commented on monetio/obs/openaq_aws.py:
```diff
@@ -0,0 +1,511 @@
+    find_paths=True,
+    n_procs=1,
+):
+    """Add OpenAQ data AWS Open Data.
```
```suggestion
    """Add OpenAQ data from AWS Open Data (https://registry.opendata.aws/openaq/).
```

zmoon commented:
> @zmoon is this ok to merge?

It shouldn't break anything, but we might want to at least add the imports like we normally do.

bbakernoaa commented:
@zmoon is this working?

zmoon commented:
> @zmoon is this working?

It was and _should_ still be. Had hoped to add the unchecked things in my PR summary, but it could be merged without that.

zmoon committed:
Include Open Data link

bbakernoaa committed:
Update environment and project configuration; add optional dependencies and remove legacy setup files

bbakernoaa committed:
Merge remote-tracking branch 'arl/develop' into pr/zmoon/237

bbakernoaa committed:
Add unit tests for AirNow and AERONET data handling; improve error handling and logging

bbakernoaa committed:
pre-commit updates and remove bad files

bbakernoaa committed:
Refactor .codespell-exclude to remove unnecessary entries and streamline the list

bbakernoaa committed:
Fix codespell action configuration to use 'ignore_words' instead of 'exclude_file'

bbakernoaa committed:
Fix codespell action configuration to use 'ignore_words_file' instead of 'ignore_words'

bbakernoaa committed:
Replace joblib with dask for parallel processing in AERONET data handling

zmoon commented:
This PR was only adding new files, so it was already mergeable without merging all of this unrelated stuff that's ahead of the develop branch in. Do you mind if I reset it to where it was?

bbakernoaa committed:
Add CI skip decorator for tests blocked by NASA requests

bbakernoaa committed:
Update environment and AERONET code to reflect transition from joblib to dask

bbakernoaa commented:
> This PR was only adding new files, so it was already mergeable without merging all of this unrelated stuff that's ahead of the develop branch in. Do you mind if I reset it to where it was?

The tests were failing locally after merging in the current develop.  I don't want to go backwards at this point as all tests pass locally including the aeronet.py when connected to the ncep vpn.

zmoon commented on .codespell-exclude:
```diff
@@ -1,42 +1,27 @@
+PRES
+RIN
+Rouge
+SIZ
+Scrip
+Wight
+fo
+hda
+indx
+metdata
+pres
```
These possible misspellings of size, weight, index, metadata, press, etc. now won't be caught. That's why the line version was better.

bbakernoaa committed:
Update documentation links and modify open_mfdataset calls to specify 'netcdf4' engine

zmoon commented on pyproject.toml:
```diff
@@ -1,14 +1,77 @@
+[tool.flake8]
+max-line-length = 130
```
Consider changing back to 100 and increasing in a separate PR. Too many diffs.

zmoon commented on pyproject.toml:
```diff
@@ -1,14 +1,77 @@
 
+[project]
+name = "monetio"
+version = "0.1.0"  # Update as needed or use dynamic versioning
```
```suggestion
dynamic = ["version"]
```

Also to specify the source: https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata

zmoon replied:
I believe still need to add later in the file
```toml
[tool.setuptools.dynamic]
version = {attr = "monetio.__version__"}
```
so it knows where to get the version from

zmoon commented on tests/test_aeronet.py:
```diff
@@ -1,3 +1,13 @@
+skip_on_ci = pytest.mark.skipif(os.environ.get("CI", "false").lower() == "true", reason="Skipped on CI")
+
+pytestmark = skip_on_ci
 from pathlib import Path
```
```suggestion

from pathlib import Path
```

zmoon reviewed (commented):
Might as well fix those 2 failing CESM tests too

bbakernoaa committed:
Update pyproject.toml

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update tests/test_aeronet.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

zmoon commented:
For the CESM-FV test, I think just need to remove this line from the fixture
https://github.com/noaa-oar-arl/monetio/blob/c19af818e4c98de992ede55910b67dd11456f790/tests/test_cesm_fv_updated.py#L39C1-L39C25
(was preventing it from doing the file lock)

bbakernoaa committed:
Enhance test file retrieval with error handling and skip logic for missing files

bbakernoaa committed:
Reduce line length settings in pre-commit configuration and pyproject.toml to 100

bbakernoaa committed:
change back to 100 linelength

bbakernoaa commented:
> For the CESM-FV test, I think just need to remove this line from the fixture https://github.com/noaa-oar-arl/monetio/blob/c19af818e4c98de992ede55910b67dd11456f790/tests/test_cesm_fv_updated.py#L39C1-L39C25 (was preventing it from doing the file lock)

its strange because it passes locally

bbakernoaa commented:
```
pytest -v tests/test_cesm_fv_updated.py 
==================================================================================== test session starts =====================================================================================
platform darwin -- Python 3.11.13, pytest-8.4.1, pluggy-1.6.0 -- /Users/l22-n04127-res/.local/share/mamba/envs/monetio-dev/bin/python3.11
cachedir: .pytest_cache
rootdir: /Users/l22-n04127-res/Documents/GitHub/monetio_repos/monetio
configfile: pyproject.toml
plugins: xdist-3.8.0, cov-6.2.1, print-1.1.0
collected 5 items                                                                                                                                                                            

tests/test_cesm_fv_updated.py::test_open_mfdataset PASSED                                                                                                                              [ 20%]
tests/test_cesm_fv_updated.py::test_open_mfdataset_surf_only_false PASSED                                                                                                              [ 40%]
tests/test_cesm_fv_updated.py::test_hybrid_vars PASSED                                                                                                                                 [ 60%]
tests/test_cesm_fv_updated.py::test_calc_pressure PASSED                                                                                                                               [ 80%]
tests/test_cesm_fv_updated.py::test_calc_pressure_i PASSED                                                                                                                             [100%]

====================================================================================== warnings summary ======================================================================================
../../../../.local/share/mamba/envs/monetio-dev/lib/python3.11/site-packages/monetio/models/_rrfs_cmaq_mm.py:5
  /Users/l22-n04127-res/.local/share/mamba/envs/monetio-dev/lib/python3.11/site-packages/monetio/models/_rrfs_cmaq_mm.py:5: DeprecationWarning: _rrfs_cmaq_mm module is deprecated. Use ufs instead.
    warnings.warn("_rrfs_cmaq_mm module is deprecated. Use ufs instead.", DeprecationWarning)

tests/test_cesm_fv_updated.py::test_open_mfdataset
  /Users/l22-n04127-res/Documents/GitHub/monetio_repos/monetio/tests/test_cesm_fv_updated.py:22: UserWarning: Downloading test file f.e22.FCnudged.f09_f09_mg17.cst_emis.cam.h1.2018-12-25-43200.nc for CESM-FV test
    warnings.warn(f"Downloading test file {fn} for CESM-FV test")

tests/test_cesm_fv_updated.py::test_open_mfdataset_surf_only_false
  /Users/l22-n04127-res/.local/share/mamba/envs/monetio-dev/lib/python3.11/site-packages/monetio/models/_cesm_fv_mm.py:44: UserWarning: 3D data processing is still experimental in CESM-FV (CAM-Chem), and has not been properly tested. Use at own risk.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============================================================================== 5 passed, 3 warnings in 28.49s ===============================================================================
(monetio-dev) l22-n04127-res@MacBook-Pro:~/.../monetio_repos/monetio$ 
```

bbakernoaa committed:
Add .vscode to .gitignore and update pyproject.toml with dynamic versioning

zmoon commented on environment-dev.yml:
```diff
@@ -31,5 +31,5 @@ dependencies:
   - pip:
-      - "-e ."
+      - " .[dev]"
       - pytest-print
```
```suggestion
```
(it can go in pyproject.toml too)

zmoon commented on pyproject.toml:
```diff
@@ -1,7 +1,72 @@
+dev = [
+    "pytest",
+    "pytest-cov",
+    "pytest-xdist",
```
```suggestion
    "pytest-print",
    "pytest-xdist",
```

zmoon commented on tests/test_cesm_fv_updated.py:
```diff
@@ -1,5 +1,7 @@
 import shutil
 import warnings
+
+# from math import e
```
```suggestion
```

zmoon commented on tests/test_cesm_fv_updated.py:
```diff
@@ -14,28 +16,31 @@
-    worker_id = "master"
+    # worker_id = "master"
 
     if worker_id == "master":
```
```suggestion
    if worker_id == "master":
```

zmoon commented on tests/test_airnow.py:
```diff
@@ -1,3 +1,5 @@
+import logging
+
```
```suggestion
```

zmoon commented on tests/test_airnow.py:
```diff
@@ -24,15 +26,18 @@ def test_build_urls():
 def test_add_data_hourly():
-    dates = pd.date_range("2021/07/01", periods=3, freq="H")
+    dates = pd.date_range("2024/07/01", periods=18, freq="h")
+    print(dates)
```
```suggestion
```

zmoon commented on tests/test_airnow.py:
```diff
@@ -24,15 +26,18 @@ def test_build_urls():
 
     # Wide format (default)
     df = airnow.add_data(dates)
+    print(df)
```
```suggestion
```

zmoon commented on tests/test_airnow.py:
```diff
@@ -24,15 +26,18 @@ def test_build_urls():
 
     # Non-wide
     df = airnow.add_data(dates, wide_fmt=False)
+    print(df)
```
```suggestion
```

zmoon commented on tests/test_airnow.py:
```diff
@@ -71,7 +76,7 @@ def test_add_data_daily():
     ],
 )
-def test_check_zero_utc_offsets(date, bad_utcoffset, request, printer):
+def test_check_zero_utc_offsets(date, bad_utcoffset, request):
```
```suggestion
def test_check_zero_utc_offsets(date, bad_utcoffset, request, printer):
```

zmoon commented on tests/test_airnow.py:
```diff
@@ -94,7 +99,7 @@ def test_check_zero_utc_offsets(date, bad_utcoffset, request, printer):
             )
             msg += bad_sites.to_string(index=False)
-            printer(msg)
+            logging.info(msg)
```
```suggestion
            printer(msg)
```
In my experience, it's more useful being able to see this every time

bbakernoaa replied:
I'd prefer to use the builtin logging

zmoon commented on tests/test_cesm_fv_updated.py:
```diff
@@ -14,28 +16,31 @@
 
 
 @pytest.fixture(scope="module")
 def test_file_path(tmp_path_factory, worker_id=None):
```
```suggestion
def test_file_path(tmp_path_factory, worker_id):
```
for consistency with the others

zmoon reviewed (commented):
CESM-FV still working for me locally in fresh env :/

zmoon commented on environment-dev.yml:
```diff
@@ -31,5 +31,5 @@ dependencies:
   - pip>=21.1
   - pip:
-      - "-e ."
+      - " .[dev]"
```
```suggestion
      - "-e .[dev]"
```
`-e` is for editable, important for development

bbakernoaa committed:
Add pytest marker for file retrieval to multiple test functions

bbakernoaa committed:
Update environment-dev.yml

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update pyproject.toml

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update tests/test_cesm_fv_updated.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update tests/test_airnow.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update tests/test_airnow.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update tests/test_cesm_fv_updated.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update tests/test_airnow.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update tests/test_cesm_fv_updated.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update tests/test_airnow.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update tests/test_airnow.py

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update environment-dev.yml

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Refactor pytest marker for file retrieval to use consistent naming

bbakernoaa committed:
Merge remote-tracking branch 'refs/remotes/zmoon/openaq_aws' into pr/zmoon/237

bbakernoaa committed:
Add pytest xdist marker for CESM-FV tests

bbakernoaa committed:
update fixture

zmoon committed:
Better to fail if there's a download issue

so we know to fix it

and if want to change this behavior should do it for all the
similar fixtures too

zmoon committed:
Back to original method

zmoon committed:
Different test file than the original

so needs a different name in the temp dir

zmoon committed:
Only one test for this file

so shouldn't need xdist group
